Translate You First Component page (#852)
reactjs/ru.react.dev: src/content/learn/your-first-component.md

---
title: Ваш первый компонент
---

<Intro>

*Компоненты* — это один из основных концептов React. Они являются основой, на которой вы строите пользовательские интерфейсы (UI), что делает их идеальным местом для начала вашего пути в React!

</Intro>

<YouWillLearn>

* Что такое компонент
* Какую роль играют компоненты в React-приложении
* Как написать ваш первый React-компонент

</YouWillLearn>

## Компоненты: строительные блоки UI {/*components-ui-building-blocks*/}

> В интернете, HTML позволяет создавать нам структурированные документы, используя встроенный набор тегов, например `<h1>` и `<li>`:

Reviewer: ```suggestion
В интернете HTML позволяет создавать нам структурированные документы, используя встроенный набор тегов, например `<h1>` и `<li>`:
```

> В Интернете, HTML позволяет нам создавать структурированные документы с использованием встроенного набора тегов, таких как `<h1>` и `<li>`:

Author: Исправил

```html
<article>

>   <h1>Мой Первый Компонент</h1>

Reviewer: ```suggestion
  <h1>Мой первый компонент</h1>
```

Кажется, это скорее англицизм писать всё с больших букв в заголовках. Здесь и ниже можно убрать)

Author: Поправил

  <ol>
    <li>Компоненты: Cтроительные Блоки UI</li>
    <li>Определение Компонента</li>
    <li>Использование Компонента</li>
  </ol>
</article>
```

> Эта разметка представляет эту статью как `<article>`, её заголовок как `<h1>` и (сокращённое) оглавление как упорядоченный список `<ol>`. Такая разметка, в сочетании с CSS для стилизации и JavaScript для создания интерактивности, лежит в основе каждой боковой панели, аватара, модального окна, выпадающего меню — каждой части UI, которую вы видите в интернете.

Reviewer 2: ```suggestion
Разметка выше представляет эту статью как `<article>`, её заголовок как `<h1>` и (сокращённое) оглавление как упорядоченный список `<ol>`. Такая разметка, в сочетании с CSS для стилизации и JavaScript для создания интерактивности, кроется в каждой боковой панели, аватарке, модальном окне, выпадающем меню — каждой части UI, которую вы видите в интернете.
```
> Эта разметка представляет эту статью

Два раза это

Author: Исправил

> React позволяет вам объединять разметку, CSS и JavaScript в пользовательские "компоненты", **переиспользуемые элементы UI для вашего приложения**. Код оглавления, который вы видели выше, можно превратить в компонент `<TableOfContents />`, который можно отрендерить на каждой странице. Внутри он всё ещё использует те же HTML-теги, такие как `<article>`, `<h1>` и т.д.

Reviewer 2: ```suggestion
С React можно объединять разметку, CSS и JavaScript в ваши собственные "компоненты", **переиспользуемые элементы UI для вашего приложения**. Код оглавления выше, можно превратить в компонент `<TableOfContents />`и отрендерить его на любой странице. Внутри он всё ещё использует те же HTML-теги, такие как `<article>`, `<h1>` и т.д.
```

> который ... который

Тяжело читается, когда 2 "который" подряд

> на **каждой** странице

На любой

Author: Исправил

> Как и с HTML-тегами, вы можете комбинировать, упорядочивать и вкладывать компоненты для создания целых страниц. Например, страница документации, которую вы сейчас читаете, состоит из React-компонентов:

Reviewer 2: ```suggestion
Как и HTML-теги, компоненты можно комбинировать, упорядочивать и вкладывать друг в друга для создания целых страниц. Например, страница документации, которую вы сейчас читаете, состоит из React-компонентов:
```

Author: Исправил

</PageLayout>
```

> По мере роста вашего проекта вы заметите, что многие из ваших дизайнов можно создать, переиспользуя уже готовые компоненты, а это ускорит разработку. Наше оглавление выше может быть добавлено на любой экран как `<TableOfContents />`! Дайте резкий старт своему проекту, используя тысячи компонентов с открытым исходным кодом, которые были созданы React-сообществом, например [Chakra UI](https://chakra-ui.com/) и [Material UI.](https://material-ui.com/)

Reviewer 2: ```suggestion
По мере роста вашего проекта вы заметите, что многие из ваших дизайнов можно создать, переиспользуя уже готовые компоненты, а это ускорит разработку. Наше оглавление выше может быть добавлено на любой экран как `<TableOfContents />`! Можно дать резкий старт своему проекту, используя тысячи компонентов с открытым исходным кодом, которые были созданы React-сообществом, например [Chakra UI](https://chakra-ui.com/) и [Material UI.](https://material-ui.com/)
```

> Дайте резкий старт своему проекту

Звучит все-таки черезчур восклицательно, мне кажется

Author: Исправил

> По мере роста вашего проекта вы заметите, что многие из ваших дизайнов могут быть составлены с помощью переиспользования уже написанных вами компонентов, ускоряя вашу разработку. Наше оглавление выше может быть добавлено на любой экран с помощью `<TableOfContents />`! Вы также можете начать свой проект, используя тысячи компонентов с открытым исходным кодом, которые были созданы React-сообществом, например [Chakra UI](https://chakra-ui.com/) и [Material UI.](https://material-ui.com/)

Reviewer 2: ```suggestion
По мере роста вашего проекта вы заметите, что многие из ваших дизайнов можно создать, переиспользуя уже готовые компоненты, а это ускорит разработку. Наше оглавление выше может быть добавлено на любой экран как `<TableOfContents />`! Дайте резкий старт своему проекту, используя тысячи компонентов с открытым исходным кодом, которые были созданы React-сообществом, например [Chakra UI](https://chakra-ui.com/) и [Material UI.](https://material-ui.com/)
```

- "По мере роста" - хорошо.
- Слишком много "ваше", "ваши" и т.д. 
- to jumpstart - не просто "начать", а "стартануть", "дать резкий старт", мы ж должны понимать в чем профит компонентного подхода

Author: Исправил

## Определение компонента {/*defining-a-component*/}

> Традиционно при создании веб-страниц разработчики размечали свой контент, а затем добавляли немного интерактивности с помощью JavaScript. Это работало отлично, когда интерактивность была просто дополнительным элементом в Интернете. Сегодня она стала обязательной для многих сайтов и всех приложений. React ставит взаимодействие на первое место, при этом используя ту же технологию: **компонент React представляет собой JavaScript-функцию, которую можно _припудрить разметкой_.** Вот как это выглядит (вы можете отредактировать пример ниже):

Reviewer 2: ```suggestion
Раньше при создании веб-страниц разработчики размечали свой контент, а затем добавляли щепотку интерактивности с помощью JavaScript. Это работало отлично, ведь интерактивность в интернете была просто приятной мелочью. Сегодня же это обязательная часть многих сайтов, еще больше - приложений. React ставит интерактивность на первое место, при этом используя ту же технологию: **React-компонент представляет собой JavaScript-функцию, которую можно _припудрить разметкой_.** Вот как это выглядит (вы можете редактировать пример ниже):
```

- "Традиционно", кажется, не подходит по стилистике, потому что речь просто о том, как было принято делать раньше, а не о каких-то древних традициях, на крайний можно заменить на "Обычно"
- "Добавляли немного интерактивности", звучит хорошо, но теряется часть семантики глагола "to sprinkle" посыпать, вкрапливать, речь может также идти про декоративную обсыпку, предлагаю "добавить щепотку интерактивности", вроде неплохо звучит
- "nice-to-have" не "дополнительный", а "приятно иметь", что-то необязательное, но приятное -> "приятная мелочь"
- "Сегодня она стала", можно добавить частицу "же", чтобы подчеркнуть констраст между прошлым и будущим 
- "React ставит взаимодействие", почему до этого interactivity это "интерактивность", а теперь "взаимодействие"? Если оставлять "взаимодействие", то нужно указать взаимодействие с чем
- "припудрить разметкой" - понравилось, можно еще "приукрасить разметкой"

</Sandpack>

А вот как создать компонент:

> ### Шаг 1: Экспортировать компонент {/*step-1-export-the-component*/}

Reviewer 2: ```suggestion
### Шаг 1: Экспорт компонента {/*step-1-export-the-component*/}
```

Reviewer 2: Да, согласен, тут же список шагов, поэтому второй вариант хуже

Префикс `export default` —  это [стандартный синтаксис JavaScript](https://developer.mozilla.org/ru/docs/web/javascript/reference/statements/export) (не является спецификой React). Он позволяет пометить основную функцию в файле, чтобы её можно было импортировать из других файлов. (Подробнее об импорте в [Импорт и Экспорт компонентов](/learn/importing-and-exporting-components)!)

> ### Шаг 2: Определить функцию {/*step-2-define-the-function*/}

Reviewer: пожалуйста добавьте пустую строку, чтобы в будущем не было конфликтов

Reviewer: здесь и ниже

Author: Добавил строку перед шагом 2. Теперь строки сошлись

Reviewer: Спасибо!

С помощью `function Profile() { }` вы определяете JavaScript-функцию с именем `Profile`.

<Pitfall>

> React-компоненты — это обычные фукнции JavaScript, но **их имена должны начинаться с заглавной буквы**, иначе они не будут работать!

Author: Исправил

</Pitfall>

### Шаг 3: Добавить разметку {/*step-3-add-markup*/}

> Компонент возвращает тег `<img />` с атрибутами `src` и `alt`. `<img />` выглядит как HTML, но на самом деле это JavaScript! Этот синтаксис называется [JSX](/learn/writing-markup-with-jsx), и он позволяет вам вставлять разметку внутрь JavaScript.

Reviewer 2: ```suggestion
Компонент возвращает тег `<img />` с атрибутами `src` и `alt`. `<img />` выглядит как HTML, но на самом деле под капотом это JavaScript! Этот синтаксис называется [JSX](/learn/writing-markup-with-jsx), и он позволяет вам вставлять разметку в JavaScript.
```

> но на самом деле это JavaScript

Потерял under the hood

Author: Исправил

Оператор `return` можно записать в одну строку, как в этом компоненте:

```js
return <img src="https://i.imgur.com/MK3eW3As.jpg" alt="Кэтрин Джонсон" />;
```

> Но если вся ваша разметка не находится на той же строке, что и ключевое слово `return`, вы должны обернуть её в пару скобок:

Reviewer 2: ```suggestion
Но если вся ваша разметка не находится на той же строке, что и ключевое слово `return`, то вы должны обернуть её в пару скобок:
```

Хочется "то" вставить

Author: Исправил

</Sandpack>

> ### Что видит барузер {/*what-the-browser-sees*/}

Reviewer 2: ```suggestion
### Что видит браузер {/*what-the-browser-sees*/}
```

Author: Исправил

Большинство React-приложений используют компоненты повсюду. Это означает, что вы будете использовать компоненты не только для переиспользуемых элементов, таких как кнопки, но также для более крупных элементов: боковых панелей, списков и в конечном итоге, целых страниц! Компоненты — это удобный способ организации кода UI и разметки, даже если некоторые из них используются только один раз.

[Фреймворки на основе React](/learn/start-a-new-react-project) пошли ещё дальше. Вместо того, чтобы использовать пустой файл HTML и позволять React "захватывать" управление страницей с помощью JavaScript, они *также* автоматически генерируют HTML из ваших React-компонентов. Это позволяет вашему приложению показывать часть контента до того, как загрузится JavaScript код.

> Тем не менее, многие веб-сайты используют React только для [добавления интерактивности на существующие HTML-страницы.](/learn/add-react-to-an-existing-project#using-react-for-a-part-of-your-existing-page) У них есть множество корневых компонентов вместо одного для всей страницы. Вы можете использовать столько функциональности React, сколько вам нужно.

Reviewer 2: ```suggestion
Тем не менее, многие веб-сайты используют React только для [добавления интерактивности на существующие HTML-страницы.](/learn/add-react-to-an-existing-project#using-react-for-a-part-of-your-existing-page) У них есть множество корневых компонентов вместо одного для всей страницы. Вы можете брать от React столько, сколько вам нужно.
```

Author: Исправил

</DeepDive>

<Recap>

> Вы только что познакомились с React! Давайте вспомним некоторые ключевые моменты.

Reviewer 2: ```suggestion
Вы только что познакомились с React! Давайте повторим некоторые ключевые моменты.
```

* React позволяет вам создавать компоненты, **переиспользуемые элементы UI для вашего приложения.**

> * В React-приложении каждый элемент UI является компонентом.

Reviewer 2: ```suggestion
* В React-приложении каждый элемент UI это компонент.
```

</Sandpack>

> Возможно, вы задаётесь вопросом, почему написание только `export` недостаточно для исправления этого примера. Вы можете узнать разницу между `export` и `export default` в разделе [Импорт и Экспорт компонентов.](/learn/importing-and-exporting-components)

Reviewer 2: ```suggestion
Возможно, вы задаётесь вопросом, почему одного только написания `export` недостаточно для того, чтобы исправить этот пример. Вы можете узнать разницу между `export` и `export default` в разделе [Импорт и Экспорт компонентов.](/learn/importing-and-exporting-components)
```

Author: Исправил. Но ИМХО как-то тоже перегруженно звучит. Вариант лучше не придумал.

> Возможно, вы задаётесь вопросом, почему одного только написания `export` недостаточно, чтобы исправить этот пример. Вы можете узнать разницу между `export` и `export default` в разделе [Импорт и Экспорт компонентов.](/learn/importing-and-exporting-components)

Reviewer 2: Так менее нагруженно, кажется

> Возможно, вы задаётесь вопросом, почему одного только написания `export` недостаточно, чтобы исправить этот пример. Вы можете узнать разницу между `export` и `export default` в разделе [Импорт и Экспорт компонентов.](/learn/importing-and-exporting-components)

Author: Исправил

</Solution>

> #### Исправьте возвращаемое выражение {/*fix-the-return-statement*/}

Reviewer 2: ```suggestion
#### Исправьте оператор return {/*fix-the-return-statement*/}
```

Author: Исправил

> Что-то не так с оператором `return`. Вы можете это исправить?

Reviewer 2: ```suggestion
Что-то не так с оператором `return`. Сможете его исправить?
```

Author: Исправил

<Hint>

> При попытке исправить это вы можете получить ошибку "Unexpected token". В этом случае убедитесь, что точка с запятой находится *после* закрывающей скобки. Оставление точки с запятой внутри `return ( )` приведёт к ошибке.

Reviewer 2: ```suggestion
При попытке исправить оператор вы можете получить ошибку "Unexpected token". В этом случае убедитесь, что точка с запятой находится *после* закрывающей скобки. Оставив точку с запятой внутри `return ( )`, вы вызовите ошибку.
```

Author: Исправил

<Solution>

> Вы можете исправить этот компонент, записав возвращаемое выражение в одну строку, например так:

Reviewer 2: ```suggestion
Вы можете исправить этот компонент, записав оператор return в одну строку вот так:
```

Author: Исправил

</Solution>

> #### Найти ошибку {/*spot-the-mistake*/}

Reviewer 2: ```suggestion
#### Найдите ошибку {/*spot-the-mistake*/}
```

Author: Исправил

> Что-то не так с объявлением и использованием компонента `Profile`. Можете найти ошибку? (Попробуйте вспомнить, как React отличает компоненты от обычных тегов HTML!)

Reviewer 2: ```suggestion
Что-то не так с тем, как объявлен и использован компонент `Profile`. Сможете найти ошибку? (Попробуйте вспомнить, как React отличает компоненты от обычных тегов HTML!)
```

Author: Исправил

> Что-то не так с тем, как объявлен и использован компонента `Profile`. Сможете найти ошибку? (Попробуйте вспомнить, как React отличает компоненты от обычных тегов HTML!)

Reviewer 2: ```suggestion
Что-то не так с тем, как объявлен и использован компонент `Profile`. Сможете найти ошибку? (Попробуйте вспомнить, как React отличает компоненты от обычных тегов HTML!)
```

Author: Исправил

<Solution>

Имена React-компонентов должны начинаться с заглавной буквы.

> Измените `function profile()` на `function Profile()`, а затем измените каждый `<profile />` на `<Profile />`:

Reviewer 2: ```suggestion
Замените `function profile()` на `function Profile()`, а затем каждый `<profile />` на `<Profile />`:
```

Author: Исправил

</Solution>

#### Ваш компонент {/*your-own-component*/}

> Напишите компонент с нуля. Вы можете дать ему любое допустимое имя и возвращать любую разметку. Если ничего не приходит на ум, то вы можете написать компонент `Congratulations`, который отображает `<h1>Хорошая работа!</h1>`. Не забудьте экспортировать компонент!

Reviewer 2: ```suggestion
Напишите компонент с нуля. Вы можете дать ему любое допустимое имя и вернуть любую разметку. Если ничего не приходит на ум, то вы можете написать компонент `Congratulations`, который отображает заголовок`<h1>Хорошая работа!</h1>`. Не забудьте экспортировать компонент!
```

> дать ... возвращать

разные формы

Author: Исправил
